components/tidb_query/src/codec/convert.rs

```diff
@@ -151,121 +83,174 @@ pub fn truncate_binary(s: &mut Vec<u8>, flen: isize) {
+        return Ok(-max_f);
     }
-    Res::Ok(f)
+    Ok(f)
 }
 
-/// Returns an overflowed error.
-macro_rules! overflow {
-    ($val:ident, $bound:tt) => {{
-        Error::Eval(
-            format!("constant {} overflows {}", $val, $bound),
-            ERR_DATA_OUT_OF_RANGE,
-        )
-    }};
+#[inline]
+fn overflow(val: impl Display, bound: impl Display) -> Error {
+    Error::Eval(
+        format!("constant {} overflows {}", val, bound),
+        ERR_DATA_OUT_OF_RANGE,
+    )
+}
+
+/// A trait for converting a value to an `Int`.
+pub trait ToInt {
+    /// Converts the given value to an `i64`
+    fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64>;
+    /// Converts the given value to an `u64`
+    fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64>;
 }
 
 impl ToInt for i64 {
+    /// This func check whether the number overflow signed upperBound and lowerBound of `tp`,
+    /// if overflow, it will handle the overflow using ctx.
+    ///
+    /// This function port from TiDB's types.ConvertIntToInt,
+    /// except that it will handle overflow using ctx,
+    /// so if you get a err returned, needn't to handle it one more time.
     fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
         let lower_bound = integer_signed_lower_bound(tp);
         // https://dev.mysql.com/doc/refman/8.0/en/out-of-range-and-overflow.html
         if *self < lower_bound {
-            ctx.handle_overflow_err(overflow!(self, lower_bound))?;
+            ctx.handle_overflow_err(overflow(self, tp))?;
             return Ok(lower_bound);
         }
         let upper_bound = integer_signed_upper_bound(tp);
         if *self > upper_bound {
-            ctx.handle_overflow_err(overflow!(self, upper_bound))?;
+            ctx.handle_overflow_err(overflow(self, tp))?;
             return Ok(upper_bound);
         }
         Ok(*self)
     }
 
+    /// This func check whether the number overflow unsigned upperBound of `tp`,
+    /// if overflow, it will handle the overflow using ctx.
+    ///
+    /// This function port from TiDB's types.ConvertIntToUint,
+    /// except that it will handle overflow using ctx,
+    /// so if you get a err returned, needn't to handle it one more time
     fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
         if *self < 0 && ctx.should_clip_to_zero() {
-            ctx.handle_overflow_err(overflow!(self, 0))?;
+            ctx.handle_overflow_err(overflow(self, tp))?;
             return Ok(0);
         }
 
         let upper_bound = integer_unsigned_upper_bound(tp);
         if *self as u64 > upper_bound {
-            ctx.handle_overflow_err(overflow!(self, upper_bound))?;
+            ctx.handle_overflow_err(overflow(self, tp))?;
             return Ok(upper_bound);
         }
         Ok(*self as u64)
     }
 }
 
 impl ToInt for u64 {
+    /// This func check whether the number overflow signed upperBound and lowerBound of `tp`,
```

Comment:
How about use following document format:
```
/// This function is ported from TiDB xxxx, which checks whether the number overflow signed upper bound and lower bound of `tp`
///
/// # Notes
/// 
/// It will handle overflow using `ctx` if overflow occurred and the caller should not handle it anymore.
```